File: 22. Generate Parentheses/22. Generate Parentheses.go
func generateParenthesis(n int) []string {
    var result []string // слайс для хранения результатов
    var backtrack func(int, int, string)
    
    backtrack = func(open, close int, sequence string) {
        if len(sequence) == n*2 { // базовый случай: если скобочная последовательность имеет нужную длину, добавляем ее в результат
            result = append(result, sequence)
            return
        }
        
        if open < n { // проверяем условие для добавления открывающей скобки
            backtrack(open+1, close, sequence+"(")
        }
        
        if close < open { // проверяем условие для добавления закрывающей скобки
            backtrack(open, close+1, sequence+")")
        }
    }
    
    backtrack(0, 0, "") // запускаем рекурсивную функцию, начинаем с 0 открытых и закрытых скобок и пустой последовательности
    
    return result
}

File: 22. Generate Parentheses/22. Generate Parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        def dfs(left_count, right_count, current_str):
            if left_count > n or right_count > left_count:
                return
            if left_count == n and right_count == n:
                result.append(current_str)
                return
            dfs(left_count + 1, right_count, current_str + '(')
            dfs(left_count, right_count + 1, current_str + ')')

        result = []
        dfs(0, 0, '')
        return result
